include/Dialect/TensorExt/IR/TensorExtCanonicalization.td

```diff
@@ -0,0 +1,101 @@
+#ifndef INCLUDE_DIALECT_TENSOREXT_IR_TENSOREXTCANONICALIZATION_TD_
+#define INCLUDE_DIALECT_TENSOREXT_IR_TENSOREXTCANONICALIZATION_TD_
+
+include "TensorExtOps.td"
+include "mlir/Dialect/Arith/IR/ArithOps.td"
+include "mlir/Dialect/Tensor/IR/TensorOps.td"
+include "mlir/IR/PatternBase.td"
+
+// TODO(#515): refactor these helpers to a common file with InsertRotate.td
+defvar DefOverflow = ConstantEnumCase<Arith_IntegerOverflowAttr, "none">;
+
+def MakeSingleResultVariadic: NativeCodeCall<"{ $0 }">;
+
+def IndexType : NativeCodeCall<"$_builder.getIndexType()">;
+
+def MakeMatchingTypedZero : NativeCodeCall<
+      "$_builder.getIntegerAttr("
+        "$_builder.getIntegerType("
+          "$0.getType().getIntOrFloatBitWidth()"
+        "), 0)">;
+
+def CreateIndexCastOp : NativeCodeCall<
+      "$_builder.create<arith::IndexCastOp>($0.getLoc(), $1.getType(), $0)">;
+
+def IsZero :
+    Constraint<
+      CPred<"llvm::cast<mlir::IntegerAttr>($0).getValue().isZero()">>;
+
+def AreOpposites :
+    Constraint<
+      CPred<"llvm::cast<mlir::IntegerAttr>($0).getValue() "
+            "== llvm::cast<mlir::IntegerAttr>($1).getValue()">>;
+
+def OutOfBoundsOfTensorDim :
+  Constraint<
+    CPred<
+      "llvm::cast<mlir::IntegerAttr>($0).getValue().getSExtValue() < 0 "
+      "|| llvm::cast<mlir::IntegerAttr>($0).getValue().getSExtValue() > "
+      "llvm::cast<mlir::ShapedType>($1.getType()).getShape()[0]"
+    >
+  >;
+
+// rotate %t, 0 -> %t
+def DropZeroRotation : Pat<
+  (TensorExt_RotateOp $tensor, (ConstantLikeMatcher APIntAttr:$c0)),
+  (replaceWithValue $tensor),
+  [(IsZero $c0)]
+>;
+
+// rotate %t, x -> rotate %t, x mod size
+def NormalizeRotationIndex : Pat<
+  (TensorExt_RotateOp $tensor, (Arith_ConstantOp:$shiftOp APIntAttr:$shiftAmount)),
+  (TensorExt_RotateOp $tensor,
+      (Arith_RemUIOp
+        $shiftOp,
+        // Only works for 1D tensors: index is taken modulo the tensor length,
```

Comment:
was this because rotation indices are variadic? (I expect a lot of the TF programs may have multi-dim tensors, for some reason inputs there are usually <1xDIM>
Can tablegen maybe do a for-loop over the dimensions, or would you just need to write a pattern for each dimension size to generalize it?

Reply:
Mainly it's because I don't understand what a rotation would mean for a multi-dimensional tensor in this context. At the very least, you'd have to specify the axes of rotation (then this could select the appropriate dimension to rotate), but that would have to correspond to a meaningful operation on the underlying ciphertext. Otherwise, we might as well use the more general upstream tensor operations like scatter/gather, or even the permutation capabilities of `vector.transfer_read/write`. We only need `tensor_ext.rotate` because of how it matches the FHE computation model, after all, and that is closely tied to how the ciphertexts are packed as coefficient arrays.

Reply:
This probably implies I should have a verifier for tensor_ext.rotate to ensure it only takes 1D inputs. Will file an issue.